frontend/hub.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+type Client struct {
+	hub *Hub
+
+	// The websocket connection.
+	conn *websocket.Conn
+
+	lock sync.Mutex
+
+	// Buffered channel of outbound messages.
+	send chan []byte
+}
+
+const (
+	pongWait = 5 * time.Second
+)
+
+type Hub struct {
+	clients map[*Client]bool
+
+	broadcast chan []byte
+
+	register chan *Client
+
+	unregister chan *Client
+}
+
+var hub = &Hub{
+	broadcast:  make(chan []byte),
+	register:   make(chan *Client),
+	unregister: make(chan *Client),
+	clients:    make(map[*Client]bool),
+}
+
+func (h *Hub) run() {
+	for {
+		select {
```

Comment:
we are not handling h.broadcast closing. (e.g. coinbase connection failed while streaming)